db/bigtable.go

```diff
@@ -861,83 +1068,89 @@ func (bigtable *Bigtable) GetValidatorAttestationHistory(validators []uint64, st
-		columnFilters = make([]gcp_bigtable.Filter, 0, len(validators))
-		for _, validator := range validators {
-			columnFilters = append(columnFilters, gcp_bigtable.ColumnFilter(fmt.Sprintf("%d", validator)))
-		}
-	}
+	filter := gcp_bigtable.LatestNFilter(1)
 
-	filter := gcp_bigtable.ChainFilters(
-		gcp_bigtable.FamilyFilter(ATTESTATIONS_FAMILY),
-		gcp_bigtable.InterleaveFilters(columnFilters...),
-	)
+	g, gCtx := errgroup.WithContext(ctx)
+	g.SetLimit(concurrency)
 
-	if len(columnFilters) == 1 { // special case to retrieve data for one validators
-		filter = gcp_bigtable.ChainFilters(
-			gcp_bigtable.FamilyFilter(ATTESTATIONS_FAMILY),
-			columnFilters[0],
-		)
-	}
-	if len(columnFilters) == 0 { // special case to retrieve data for all validators
-		filter = gcp_bigtable.FamilyFilter(ATTESTATIONS_FAMILY)
-	}
-	err = bigtable.tableBeaconchain.ReadRows(ctx, ranges, func(r gcp_bigtable.Row) bool {
-		keySplit := strings.Split(r.Key(), ":")
+	for i := 0; i < len(validators); i += batchSize {
 
-		attesterSlot, err := strconv.ParseUint(keySplit[4], 10, 64)
-		if err != nil {
-			logger.Errorf("error parsing slot from row key %v: %v", r.Key(), err)
-			return false
+		upperBound := i + batchSize
+		if len(validators) < upperBound {
+			upperBound = len(validators)
 		}
-		attesterSlot = max_block_number - attesterSlot
-		for _, ri := range r[ATTESTATIONS_FAMILY] {
-			inclusionSlot := max_block_number - uint64(ri.Timestamp)/1000
+		vals := validators[i:upperBound]
 
-			status := uint64(1)
-			if inclusionSlot == max_block_number {
-				inclusionSlot = 0
-				status = 0
-			} else if orphanedSlotsMap[inclusionSlot] {
-				status = 0
-			}
-
-			validator, err := strconv.ParseUint(strings.TrimPrefix(ri.Column, ATTESTATIONS_FAMILY+":"), 10, 64)
-			if err != nil {
-				logger.Errorf("error parsing validator from column key %v: %v", ri.Column, err)
-				return false
+		g.Go(func() error {
+			select {
+			case <-gCtx.Done():
+				return gCtx.Err()
+			default:
 			}
+			ranges := bigtable.getValidatorsEpochRanges(vals, startEpoch, endEpoch)
+			err = bigtable.tableValidatorAttestations.ReadRows(ctx, ranges, func(r gcp_bigtable.Row) bool {
+				keySplit := strings.Split(r.Key(), ":")
 
-			if res[validator] == nil {
-				res[validator] = make([]*types.ValidatorAttestation, 0)
-			}
+				validator, err := bigtable.validatorKeyToIndex(keySplit[1])
+				if err != nil {
+					logger.Errorf("error parsing validator from row key %v: %v", r.Key(), err)
+					return false
+				}
+
+				for _, ri := range r[ATTESTATIONS_FAMILY] {
+					attesterSlotString := strings.Replace(ri.Column, ATTESTATIONS_FAMILY+":", "", 1)
+					attesterSlot, err := strconv.ParseUint(attesterSlotString, 10, 64)
+					if err != nil {
+						logger.Errorf("error parsing slot from row key %v: %v", r.Key(), err)
+						return false
+					}
+					inclusionSlot := MAX_BLOCK_NUMBER - uint64(ri.Timestamp)/1000
+
+					status := uint64(1)
+					if inclusionSlot == MAX_BLOCK_NUMBER {
```

Comment:
I don't think this check is necessary. We save ri.Timestamp with `(MAX_BLOCK_NUMBER-inclusionSlot)*1000` so (lets assume MAX_BLOCK_NUMBER 99) 99-0=99. *1000 = 99000
And convert back to inclusionSlot by `MAX_BLOCK_NUMBER - uint64(ri.Timestamp)/1000`. So 99 - (99000)/1000 = 99 - 99 = 0.
This case only handles when we store inclusionSlot `MAX_BLOCK_NUMBER` but then we shouldn't manually correct the inclusion slot to 0.